File: angleAverage/angle-average.c
#include <stdio.h>
#include <math.h>

int main(){
  double a[10]={347,350,352,354,356,0,0,0,0,0};
  double b[10]={270,280,290,300,0,0,0,0,0,0};
  double aver=0,aver_c,aver_s,sum_sin=0,sum_cos=0,mid;

  int i=0,n=4; // average of how many numbers

  for(i=0;i<4;i++){ // average of how many numbers
    mid = b[i]*M_PI/180; // change the degrees to radians
    sum_sin += sin(mid); // sum of sin values of all angles
    sum_cos += cos(mid); // sum of cos values of all angles
  }
  aver_s = sum_sin/n; aver_c = sum_cos/n; // average of sin and cos
  // the code below is to calculate the tan value of the average angles
  if (aver_c < 0)
    aver = atan(aver_s/aver_c)*180/M_PI+180;
  else if (aver_s > 0)
    aver = atan(aver_s/aver_c)*180/M_PI;
  else if (aver_s <= 0)
    aver = atan(aver_s/aver_c)*180/M_PI+360;

  printf("\naverage:%f\n\n", aver);
  return 0;
}

File: angleAverage/angleAverage.js
// [347, 350, 352, 354, 356, 0, 0, 0, 0, 0, ]
// result 355.903214

// [352, 350, 0, 0, 0, 0, 0, 0, 0, 0,]

const a = [347, 350, 352, 354, 356, 0, 0, 0, 0, 0,];
const b = [270, 280, 290, 300, ]; // 0, 0, 0, 0, 0, 0,

let aver = 0;
let aver_c;
let aver_s;
let sum_sin = 0;
let sum_cos = 0;
let mid;

for (let i = 0; i < b.length; i++) {
  mid = b[i] * Math.PI / 180;
  sum_sin += Math.sin(mid);
  sum_cos += Math.cos(mid);
}

aver_s = sum_sin / b.length;
aver_c = sum_cos / b.length;

if (aver_c < 0) {
  aver = Math.atan(aver_s / aver_c) * 180 / Math.PI + 180;
} else if (aver_s > 0) {
  aver = Math.atan(aver_s / aver_c) * 180 / Math.PI;
} else if (aver_s <= 0) {
  aver = Math.atan(aver_s / aver_c) * 180 / Math.PI + 360;
}

console.log(aver);




const angles = [270, 280, 290, 300, ];

const toDegrees = (angle) => {
  return angle * (180 / Math.PI);
};

const toRadians = (angle) => {
  return angle * (Math.PI / 180);
};

const angleAverage = (angles) => {
  let sumSin = 0;
  let sumCos = 0;

  angles.forEach(angle => {
    const angleRadian = toRadians(angle);
    sumSin += Math.sin(angleRadian);
    sumCos += Math.cos(angleRadian);
  });

  const avgSin = sumSin / angles.length;
  const avgCos = sumCos / angles.length;

  if (avgCos < 0) return toDegrees(Math.atan(avgSin / avgCos)) + 180;
  else if (avgSin > 0) return toDegrees(Math.atan(avgSin / avgCos));
  else if (avgSin <= 0) return toDegrees(Math.atan(avgSin / avgCos)) + 360;
};

console.log(angleAverage(angles));
